src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/measure/ActionMeter.kt

```diff
@@ -76,50 +109,82 @@ class ActionMeter(
         observation: (T) -> JsonObject?
     ): T {
         val start = clock.instant()
+        val actionMetricBuilder = ActionMetric.Builder(
+            label = key.label,
+            virtualUser = virtualUser,
+            start = start
+        )
+
         try {
             val result = action()
-            output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = ActionResult.OK,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
-                    .observation(observation(result))
-                    .drilldown(w3cPerformanceTimeline.record())
-                    .build()
-            )
+            result?.let { actionMetricBuilder.observation(observation(result)) }
+            actionMetricBuilder.result(ActionResult.OK)
             return result
         } catch (e: Exception) {
-            val result = if (e.representsInterrupt()) {
+            val actionResult = if (e.representsInterrupt()) {
                 ActionResult.INTERRUPTED
             } else {
                 ActionResult.ERROR
             }
+            actionMetricBuilder.result(actionResult)
+            throw Exception("Action '${key.label}' $actionResult", e)
+        } finally {
+            actionMetricBuilder.duration(Duration.between(start, clock.instant()))
+
             output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = result,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
-                    .drilldown(w3cPerformanceTimeline.record())
-                    .build()
+                actionMetricBuilder.build().let {
+                    if (drilldownCondition.test(it)) {
+                        actionMetricBuilder
+                            .drilldown(w3cPerformanceTimeline.record())
+                            .build()
+                    } else {
+                        it
+                    }
+                }
             )
-            throw Exception("Action '${key.label}' $result", e)
         }
     }
 
+    @Deprecated(message = "Use ActionMeter.Builder instead")
     fun withW3cPerformanceTimeline(
         w3cPerformanceTimeline: W3cPerformanceTimeline
     ): ActionMeter = ActionMeter(
         virtualUser = virtualUser,
         output = output,
         clock = clock,
-        w3cPerformanceTimeline = w3cPerformanceTimeline
+        w3cPerformanceTimeline = w3cPerformanceTimeline,
+        drilldownCondition = Predicate { true }
     )
+
+    class Builder(
+        private val virtualUser: UUID,
+        private val output: ActionMetricOutput
```

Comment:
A default of `ThrowawayActionMetricOutput` could work. Such a throwaway meter is useful sometimes, but I guess it could be a big WTF factor if by default the meter would drop measurements. So I agree, the choice of `output` should be forced on the caller 👍